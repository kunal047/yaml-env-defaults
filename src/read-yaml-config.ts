import { safeLoad } from 'js-yaml';
import { set, isPlainObject } from 'lodash';
import { readFileSync } from 'fs';

type EnvVariables = Record<string, any>;

function replaceEnvVarRefsWithDefaults(val: string, customEnv?: EnvVariables) {
  const foundValue = val.replace(/\$\{(\w+)\}/g, (substring, envVarName) => {
    const mergedValues = customEnv ? {
      ...process.env,
      ...customEnv
    } : process.env;

    const envVarValue = mergedValues[envVarName];
    if (envVarValue !== undefined) {
      return envVarValue;
    }
    throw new Error(
      `Unknown environment variable referenced in config: ${envVarName}. Specify environment variable or use defaults syntax.`
    );
  });
  return foundValue.replace(
    /(.*)\${(\w+):(.*)}(.*)/g,
    (match: string, prefix, envVarName, defValue, postfix) => {
      const envValue = process.env[envVarName];
      if (envValue !== undefined) {
        return `${prefix}${envValue.trim()}${postfix}`;
      } else {
        if (prefix || postfix) {
          return transformIfJSON(`${prefix}${transformIfJSON(defValue)}${postfix}`);
        } else {
          return transformIfJSON(defValue);
        }
      }
    }
  );
}

function transformIfJSON(val: string) {
  try {
    return JSON.parse(val);
  } catch (e) {
    return val;
  }
}

function processYaml(
  target: Record<string, any>,
  value: any,
  key?: string,
  envVariables?: EnvVariables,
): Record<string, any> {
  if ((isPlainObject(value) || Array.isArray(value)) && Object.keys(value).length > 0) {
    for (const [subKey, subVal] of Object.entries(value)) {
      processYaml(target, subVal, key ? `${key}.${subKey}` : subKey, envVariables);
    }
  } else if (key !== undefined) {
    set(target, key, typeof value === 'string' ? replaceEnvVarRefsWithDefaults(value, envVariables) : value);
  }

  return target;
}

/**
 * Replace env refs in yaml to values
 * @param {string | string[]} filePath path to yaml file
 * @param {Object<*>} customEnv custom environment object
 * @returns {Record<string, any>}
 */
export const readYamlEnv = (filePath: string | string[], customEnv?: Record<string, any>): Record<string, any> => {
  let replacedYaml = {};
  if (Array.isArray(filePath)) {
    for (const file of filePath) {
      const yaml = safeLoad(readFileSync(file, 'utf-8'));
      if (yaml) {
        replacedYaml = processYaml(replacedYaml, yaml, undefined, customEnv);
      }
    }
  } else {
    const yaml = safeLoad(readFileSync(filePath, 'utf-8'));
    if (yaml) {
      replacedYaml = processYaml(replacedYaml, yaml, undefined, customEnv);
    }
  }
  return replacedYaml;
};
