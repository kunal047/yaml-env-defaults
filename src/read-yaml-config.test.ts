import { resolve } from 'path';
import { readYamlEnv } from './read-yaml-config';

const getFixtureFile = (name: string) => resolve(__dirname, '__fixtures__', name);

describe('read yaml config with env replacing', () => {
  test('reads single yaml from file system and parses to json', () => {
    const config = readYamlEnv(getFixtureFile('config.yml'));

    expect(config).toMatchSnapshot();
  });

  test('returns a deep object from flat yaml file', () => {
    const config = readYamlEnv(getFixtureFile('flat_config.yml'));

    expect(config).toMatchSnapshot();
  });

  test('merges two yaml config files', () => {
    const config = readYamlEnv([getFixtureFile('config.yml'), getFixtureFile('flat_config.yml')]);
    expect(config).toMatchSnapshot();
  });

  test('should inject an environment variable value when setting a value with ${ENV_VAR}', () => {
    process.env.ENV_VAR_MY = 'Some value 1';
    process.env.ENV_VAR_YOUR = 'Another Value 2';

    const config = readYamlEnv(getFixtureFile('config-env.yml'));

    delete process.env.ENV_VAR_MY;
    delete process.env.ENV_VAR_YOUR;

    expect(config).toMatchSnapshot();
  });

  test('should read config with defaults env values', () => {
    const config = readYamlEnv(getFixtureFile('config-env-defaults.yaml'));

    expect(config).toMatchSnapshot();
  });

  test('fail when environment variable is not specified', () => {
    expect(() => {
      readYamlEnv([getFixtureFile('/config-env.yml')]);
    }).toThrow();
  });

  describe('.readYamlEnv() with custom env object', () => {

    test('should read config with customEnvs when process.env is empty', () => {
      const config = readYamlEnv([getFixtureFile('./config-env.yml')], {
        ENV_VAR_YOUR: 'Another Value',
        ENV_VAR_MY: 'Some value 2'
      });

      expect(config).toMatchSnapshot();
    });

    test('should replace to custom env value instead of process.env value', () => {
      process.env.ENV_VAR_MY = 'PROCESS_ENV';
      const config = readYamlEnv([getFixtureFile('./config-env.yml')], {
        ENV_VAR_YOUR: 'Another Value',
        ENV_VAR_MY: 'CUSTOM_ENV'
      });

      expect(config.someStrange).toMatch(/CUSTOM_ENV/);

      expect(config).toMatchSnapshot();
    });

    test('should not mutate customEnv', () => {
      const customEnv = {
        ENV_VAR_YOUR: 'Another Value',
        ENV_VAR_MY: 'CUSTOM_ENV'
      };
      const copy = JSON.stringify(customEnv);
      readYamlEnv([getFixtureFile('./config-env.yml')], customEnv);

      expect(copy).toEqual(JSON.stringify(customEnv));
    });
  });
});

