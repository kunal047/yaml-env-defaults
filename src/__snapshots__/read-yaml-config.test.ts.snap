// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merges two yaml config files 1`] = `
Object {
  "app": Object {
    "arr": Array [
      1,
    ],
    "array": Array [
      1,
      2,
    ],
    "boolean": true,
    "empty-array": Array [],
    "empty-object": Object {},
    "empty_arr": Array [],
    "empty_obj": Object {},
    "falseValue": false,
    "obj": Object {
      "val": 3,
    },
    "object": Object {
      "val": 3,
    },
    "text": "some text",
    "trueValue": true,
    "yaml-array": Array [
      "first",
      "second",
      "third",
    ],
  },
  "arr": Array [
    1,
    2,
  ],
  "empty_arr": Array [],
  "empty_obj": Object {},
  "obj": Object {
    "val": 3,
  },
}
`;

exports[`read yaml config with env replacing .readYamlEnv() with custom env object should read config with customEnvs when process.env is empty 1`] = `
Object {
  "bar": "text-Some value 2-text-Another Value-text",
  "escapedValue": Object {
    "value": "\\\\$\\\\{ENV_VAR_MY\\\\}",
  },
  "foo": 1,
  "someFoo": Object {
    "someBar": Array [
      "Some value 2",
      "Another Value",
    ],
    "yaml-array": Array [
      "Some value 2",
      "second",
    ],
  },
  "someStrange": "Thats/Some value 2/cloud-or-not/Another Value",
}
`;

exports[`read yaml config with env replacing .readYamlEnv() with custom env object should replace to custom env value instead of process.env value 1`] = `
Object {
  "bar": "text-CUSTOM_ENV-text-Another Value-text",
  "escapedValue": Object {
    "value": "\\\\$\\\\{ENV_VAR_MY\\\\}",
  },
  "foo": 1,
  "someFoo": Object {
    "someBar": Array [
      "CUSTOM_ENV",
      "Another Value",
    ],
    "yaml-array": Array [
      "CUSTOM_ENV",
      "second",
    ],
  },
  "someStrange": "Thats/CUSTOM_ENV/cloud-or-not/Another Value",
}
`;

exports[`read yaml config with env replacing merges two yaml config files 1`] = `
Object {
  "app": Object {
    "arr": Array [
      1,
    ],
    "array": Array [
      1,
      2,
    ],
    "boolean": true,
    "empty-array": Array [],
    "empty-object": Object {},
    "empty_arr": Array [],
    "empty_obj": Object {},
    "falseValue": false,
    "obj": Object {
      "val": 3,
    },
    "object": Object {
      "val": 3,
    },
    "text": "some text",
    "trueValue": true,
    "yaml-array": Array [
      "first",
      "second",
      "third",
    ],
  },
  "arr": Array [
    1,
    2,
  ],
  "empty_arr": Array [],
  "empty_obj": Object {},
  "obj": Object {
    "val": 3,
  },
}
`;

exports[`read yaml config with env replacing reads single yaml from file system and parses to json 1`] = `
Object {
  "app": Object {
    "arr": Array [
      1,
    ],
    "empty_arr": Array [],
    "empty_obj": Object {},
    "falseValue": false,
    "obj": Object {
      "val": 3,
    },
    "text": "some text",
    "trueValue": true,
    "yaml-array": Array [
      "first",
      "second",
      "third",
    ],
  },
  "arr": Array [
    1,
    2,
  ],
  "empty_arr": Array [],
  "empty_obj": Object {},
  "obj": Object {
    "val": 3,
  },
}
`;

exports[`read yaml config with env replacing returns a deep object from flat yaml file 1`] = `
Object {
  "app": Object {
    "array": Array [
      1,
      2,
    ],
    "boolean": true,
    "empty-array": Array [],
    "empty-object": Object {},
    "object": Object {
      "val": 3,
    },
    "text": "some text",
  },
}
`;

exports[`read yaml config with env replacing should inject an environment variable value when setting a value with \${ENV_VAR} 1`] = `
Object {
  "bar": "text-Some value 1-text-Another Value 2-text",
  "escapedValue": Object {
    "value": "\\\\$\\\\{ENV_VAR_MY\\\\}",
  },
  "foo": 1,
  "someFoo": Object {
    "someBar": Array [
      "Some value 1",
      "Another Value 2",
    ],
    "yaml-array": Array [
      "Some value 1",
      "second",
    ],
  },
  "someStrange": "Thats/Some value 1/cloud-or-not/Another Value 2",
}
`;

exports[`read yaml config with env replacing should read config with defaults env values 1`] = `
Object {
  "some": Object {
    "params": Object {
      "boolean": "false",
      "long": "12.3",
      "number": "1234",
      "test": "default value",
    },
  },
}
`;

exports[`reads single yaml from file system and parses to json 1`] = `
Object {
  "app": Object {
    "arr": Array [
      1,
    ],
    "empty_arr": Array [],
    "empty_obj": Object {},
    "falseValue": false,
    "obj": Object {
      "val": 3,
    },
    "text": "some text",
    "trueValue": true,
    "yaml-array": Array [
      "first",
      "second",
      "third",
    ],
  },
  "arr": Array [
    1,
    2,
  ],
  "empty_arr": Array [],
  "empty_obj": Object {},
  "obj": Object {
    "val": 3,
  },
}
`;

exports[`returns a deep object from flat yaml file 1`] = `
Object {
  "app": Object {
    "array": Array [
      1,
      2,
    ],
    "boolean": true,
    "empty-array": Array [],
    "empty-object": Object {},
    "object": Object {
      "val": 3,
    },
    "text": "some text",
  },
}
`;

exports[`should inject an environment variable value when setting a value with \${ENV_VAR} 1`] = `
Object {
  "bar": "text-Some value 1-text-Another Value 2-text",
  "escapedValue": Object {
    "value": "\\\\$\\\\{ENV_VAR_MY\\\\}",
  },
  "foo": 1,
  "someFoo": Object {
    "someBar": Array [
      "Some value 1",
      "Another Value 2",
    ],
    "yaml-array": Array [
      "Some value 1",
      "second",
    ],
  },
  "someStrange": "Thats/Some value 1/cloud-or-not/Another Value 2",
}
`;

exports[`should read config with defaults env values 1`] = `
Object {
  "some": Object {
    "params": Object {
      "boolean": "false",
      "long": "12.3",
      "number": "1234",
      "test": "default value",
    },
  },
}
`;
